Code/GraphMol/ChemReactions/Wrap/testEnumerations.py

```diff
@@ -0,0 +1,433 @@
+#       permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+from __future__ import print_function
+
+import unittest
+import os,sys, copy
+
+from rdkit.six.moves import cPickle
+
+from rdkit import rdBase
+from rdkit import Chem
+from rdkit.Chem import rdChemReactions
+from rdkit import Geometry
+from rdkit import RDConfig
+import itertools, time
+
+def log(s):
+  Chem.LogErrorMsg("== " + s)
+  
+class TestCase(unittest.TestCase) :
+  def setUp(self):
+    self.dataDir = os.path.join(RDConfig.RDBaseDir,'Code','GraphMol','ChemReactions','testData')
+
+  def testCartesianProduct(self):
+    log("testCartesianProduct")
+    rxn = rdChemReactions.ChemicalReaction();
+    rgroups = [[Chem.MolFromSmiles("C")]*10,
+               [Chem.MolFromSmiles("N")]*5,
+               [Chem.MolFromSmiles("O")]*6]
+
+    cartProd = rdChemReactions.CartesianProductStrategy()
+    cartProd.Initialize(rxn, rgroups)
+    self.assertEquals(cartProd.GetNumPermutations(), 10*5*6)
+    groups = []
+    count = 0
+    print (cartProd.__bool__())
+    while cartProd:
+      groups.append(tuple(cartProd.next()))
+#      count += 1
+#      assert count <= cartProd.GetNumPermutations()
+    self.assertEquals(len(groups), 10*5*6)
+    # see if we are equal to the Python implementation
+    g = list(itertools.product( list(range(10)), list(range(5)), list(range(6)) ))
+    self.assertEquals(set(g), set(groups))
+    copy.copy(cartProd)
+    
+  def testRandomSample(self):
+    log("testRandomSample")
+    rgroups = [[Chem.MolFromSmiles("C")]*10,
+               [Chem.MolFromSmiles("N")]*5,
+               [Chem.MolFromSmiles("O")]*6]
+    rxn = rdChemReactions.ChemicalReaction();
+
+    randProd = rdChemReactions.RandomSampleStrategy()
+    randProd.Initialize(rxn, rgroups)
+    self.assertEquals(randProd.GetNumPermutations(), 10*5*6)
+    groups = []
+    for i in range(10*5*6):
+      groups.append(tuple(randProd.next()))
+    print( len(set(groups)), "out of", 10*5*6 )
+
+    randProd = rdChemReactions.RandomSampleStrategy()
+    randProd.Initialize(rxn, rgroups)
+    self.assertEquals(randProd.GetNumPermutations(), 10*5*6)
+    groups = []
+    for i in range(10):
+      groups.append(tuple(randProd.next()))
+
+    for i in range(3):
```

Comment:
Again, this is really testing a lot. It would be good to check the actual results.